### DISCUSSION TITLE
How to use Ranger optimizer with YOLOV8?

### DISCUSSION THREAD

**Discussion by triciardvn** [ultralytics/ultralytics#16410] [General]
This is the error for my code since I want to try the "Ranger" optimizer instead of the existing available ones using YOLOV8. 

`NotImplementedError: Optimizer 'Ranger' not found in list of available optimizers [Adam, AdamW, NAdam, RAdam, RMSProp, SGD, auto].To request support for addition optimizers please visit https://github.com/ultralytics/ultralytics.`

How can I integrate this new optimizer as I want to do an object detection for etrikes and ebikes (I want to start training it). I already tried this code:
```
!pip install git+https://github.com/lessw2020/Ranger21.git

from ultralytics import YOLO
import os

# Load your model
model = YOLO('yolov8n.pt')

# Define hyperparameters
hyperparameters = {
    'lr': 0.001,
    'weight_decay': 0.0005,
    'momentum': 0.937
}

# Get the number of training images
num_train_images = len(os.listdir('/content/drive/MyDrive/e/train/images'))
batch_size = 16
num_batches_per_epoch = num_train_images // batch_size
num_epochs = 10

# Setup training
model.train(data='/content/drive/MyDrive/e/data.yaml', epochs=num_epochs, imgsz=640, batch=batch_size, optimizer='Ranger')
```

Please let me know what to do ASAP. Thank you. 

Note: I need to use the ranger optimizer as I need it for my thesis.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @triciardvn, thank you for reaching out to the Ultralytics community 🚀! We're thrilled to assist you. 

It seems you're experiencing an issue integrating the "Ranger" optimizer with YOLOv8. As this is a feature request or customization, an Ultralytics engineer will review your case soon. In the meantime, I recommend looking through our [Docs](https://docs.ultralytics.com) for tips and existing functionalities that might help you achieve your goals.

For 🐛 Bug Reports, please ensure you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to aid in debugging.

Feel free to share more details about your setup and any specific errors you encounter. If you've followed our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/), it might also be helpful to provide any relevant logs or data snippets.

Stay connected with the community through our channels. For real-time support, hop into our [Discord](https://ultralytics.com/discord) 🎧. Prefer detailed threads? Join discussions on [Discourse](https://community.ultralytics.com) or explore insights on [Reddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure your `ultralytics` package is updated along with all required [dependencies](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

You can run YOLOv8 using various up-to-date verified environments. Check out:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a> 

Check the CI status to ensure all YOLOv8 [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) are operating well.

Your feedback and contributions make a big difference. Keep us posted, and you'll have more updates from our team soon! 😊

**Comment by glenn-jocher** [comment]
@triciardvn currently, YOLOv8 does not support the Ranger optimizer directly. You might consider modifying the source code to integrate it, but this would require a deep understanding of the codebase. For now, using one of the supported optimizers is recommended.
